// reacty/index.js
'use strict'

var React  = require('react')
var F      = require('functionally')
var classy = require('classy')

var prepareMixin = F.curry(function(ReactyClass, mixin){
    classy.mixin(ReactyClass.prototype, mixin)
})

function prepareMixins(ReactyClass, mixins){
    if (mixins && mixins.length){
        var prepare = prepareMixin(ReactyClass)
        // debugger
        mixins.forEach(prepare)
    }
}

function getInfo(target){
    var ReactDescriptor = target
    var ReactyClass
    var ReactClass

    if (target.ReactDescriptor && target.ReactClass){
        ReactyClass = target
        ReactDescriptor = target.ReactDescriptor
    } else {
        //target is a react descriptor
        ReactyClass = target._reacty
    }

    ReactClass = ReactDescriptor.type

    return {
        ReactClass     : ReactClass,
        ReactDescriptor: ReactDescriptor,
        ReactyClass    : ReactyClass
    }
}

var Reacty = {
    define: function(config){

        var mixins = config.mixins
        delete config.mixins

        var CLASS = classy.define(config)
        delete CLASS.prototype.constructor

        prepareMixins(CLASS, mixins)

        var ReactDescriptor = React.createClass(CLASS.prototype)
        var ReactClass = ReactDescriptor.type

        CLASS.ReactDescriptor = ReactDescriptor
        CLASS.ReactClass = ReactDescriptor

        ReactDescriptor._reacty = CLASS
        ReactClass._reacty      = CLASS

        return ReactDescriptor
    },

    override: function(target, config){
        var info = getInfo(target)

        classy.core.overrideObject(info.ReactClass.prototype, config)
        // info.ReactyClass.override(config)
    }
}

module.exports = Reacty
